components/camel-graphql/src/main/java/org/apache/camel/component/graphql/GraphqlEndpoint.java

```diff
@@ -0,0 +1,225 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.graphql;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+
+import org.apache.camel.Component;
+import org.apache.camel.Consumer;
+import org.apache.camel.Processor;
+import org.apache.camel.Producer;
+import org.apache.camel.RuntimeCamelException;
+import org.apache.camel.spi.Metadata;
+import org.apache.camel.spi.UriEndpoint;
+import org.apache.camel.spi.UriParam;
+import org.apache.camel.spi.UriPath;
+import org.apache.camel.support.DefaultEndpoint;
+import org.apache.camel.util.IOHelper;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.camel.util.json.JsonObject;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpHost;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.UsernamePasswordCredentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.utils.HttpClientUtils;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
+
+@UriEndpoint(firstVersion = "3.0.0", scheme = "graphql", title = "GraphQL", syntax = "graphql:httpUri?options", label = "graphql", producerOnly = true)
+public class GraphqlEndpoint extends DefaultEndpoint {
+
+    @UriPath
+    @Metadata(required = true)
+    private URI httpUri;
+    @UriParam
+    private String proxyHost;
+    @UriParam
+    private String accessToken;
+    @UriParam
+    private String username;
+    @UriParam
+    private String password;
+    @UriParam
+    private String query;
+    @UriParam
+    private String queryFile;
+    @UriParam
+    private String operationName;
+    @UriParam
+    private JsonObject variables;
+
+    private CloseableHttpClient httpClient;
+
+    public GraphqlEndpoint(String uri, Component component) {
+        super(uri, component);
+    }
+
+    @Override
+    protected void doStop() throws Exception {
+        HttpClientUtils.closeQuietly(this.httpClient);
+    }
+
+    @Override
+    public Producer createProducer() throws Exception {
+        return new GraphqlProducer(this);
+    }
+
+    @Override
+    public Consumer createConsumer(Processor processor) throws Exception {
+        throw new UnsupportedOperationException("You cannot receive messages at this endpoint: " + getEndpointUri());
+    }
+
+    public CloseableHttpClient getHttpclient() {
```

Comment:
Its better to do all those initialization in doStart, eg also the one that loads the query file, then all initialization is done at the right moment

Reply:
Thanks for the code review. I will move the HttpClient creation and query file load to doStart as part of a future enhancement/bug fix.